### PR TITLE
Version mismatch #244

### LINKED ISSUE #244: Warn on version mismatches
Reported by @ahayschi and @reidmcy

We should check for version mismatches and raise an error under certain circumstances. Related to #228.

### PR DESCRIPTION
**Pull request by yadudoc**
Fixes #244

### CODE REVIEW

File: parsl/executors/high_throughput/process_worker_pool.py [hunk 1]
```diff
@@ -149,6 +174,7 @@ def pull_tasks(self, kill_event):
                         #    [i['task_id'] for i in self.pending_task_queue]))
             else:
                 logger.debug("[TASK_PULL_THREAD] No incoming tasks")
+                poll_timer = poll_timer * 2
```
Comment (benclifford) [inline parsl/executors/high_throughput/process_worker_pool.py]: if i have a workflow that has an initial stage that takes 1h and doesn't do much, and then generates 1000 smaller tasks, and i allocated 1000 nodes, does this mean that 999 workers might wait "a long time" (~1h?) after the 1000 smaller tasks become ready before they begin work, due to poll_timer getting very big over that initial hour?

File: parsl/executors/high_throughput/process_worker_pool.py [hunk 2]
```diff
@@ -275,10 +301,12 @@ def worker(worker_id, pool_id, task_queue, result_queue, worker_queue):
     """
     start_file_logger('{}/{}/worker_{}.log'.format(args.logdir, pool_id, worker_id),
                       worker_id,
-                      level=logging.DEBUG if args.debug is True else logging.INFO)
+                      name="worker_log",
+                      level=logging.DEBUG if args.debug else logging.INFO)
 
     # Sync worker with master
     logger.info('Worker {} started'.format(worker_id))
+    logger.debug("Debug logging enabled")
```
Comment (benclifford) [inline parsl/executors/high_throughput/process_worker_pool.py]: only output this if args.debug - there might be other third party loggers in place, and they'll show a "debug logging enabled" message even when it isn't.

File: parsl/executors/extreme_scale/mpi_worker_pool.py [hunk 3]
```diff
@@ -44,31 +47,44 @@ def __init__(self,
         worker_url : str
              Worker url on which workers will attempt to connect back
         """
-        logger.info("Manager started v0.5")
+        logger.info("Manager started v0.6")
```
[the remark this answers is not in the document]
Comment (yadudoc) [reply parsl/executors/extreme_scale/mpi_worker_pool.py]: This was purely for debugging. Removed.

File: parsl/executors/high_throughput/interchange.py [hunk 4]
```diff
@@ -3,15 +3,19 @@
...
 import logging
 import queue
 import threading
+import json
 
+from parsl.version import VERSION as PARSL_VERSION
 from ipyparallel.serialize import serialize_object
 
-LOOP_SLOWDOWN = 0.0  # in seconds
+LOOP_SLOWDoOWN = 0.0  # in seconds
```
[the remark this answers is not in the document]
Comment (yadudoc) [reply parsl/executors/high_throughput/interchange.py]: fixed

File: parsl/executors/high_throughput/process_worker_pool.py (hunk 1 above)
Comment (yadudoc) [reply parsl/executors/high_throughput/process_worker_pool.py]: Good catch. I've added a condition to limit timeout increase to self.heartbeat_period

File: parsl/executors/high_throughput/process_worker_pool.py [hunk 5]
```diff
@@ -388,6 +419,13 @@ def set_stream_logger(name='parsl', level=logging.DEBUG, format_string=None):
                           level=logging.DEBUG if args.debug is True else logging.INFO)
 
         logger.info("Python version: {}".format(sys.version))
+        logger.info("Debug logging?: {}".format(args.debug))
```
Comment (annawoodard) [inline parsl/executors/high_throughput/process_worker_pool.py]: ?: -> :

File: parsl/executors/high_throughput/interchange.py [hunk 6]
```diff
@@ -263,26 +294,39 @@ def start(self, poll_period=1):
...
-                                                          # 'wtime': 60,
-                                                          'free_capacity': tasks_requested,
+                                                          'free_capacity': 0,
+                                                          'active': True,
                                                           'tasks': []}
+                    self._ready_manager_queue[manager].update(msg)
+                    logger.info("Registration info for manager {}: {}".format(manager, msg))
+                    if (msg['python_v'] != self.current_platform['python_v'] or
+                        msg['parsl_v'] != self.current_platform['parsl_v']):
+                        logger.warn("Manager:{} has incompatible version info with the interchange".format(manager))
```
[the remark this answers is not in the document]
Comment (yadudoc) [reply parsl/executors/high_throughput/interchange.py]: Fixed

File: parsl/executors/extreme_scale/mpi_worker_pool.py [hunk 7]
```diff
@@ -44,31 +47,46 @@ def __init__(self,
...
-        self.ready_worker_queue = queue.Queue(maxsize=max_queue_size + 10)
+        self.ready_worker_queue = queue.Queue(maxsize=max_queue_size + 10 ^ 3)
 
         self.tasks_per_round = 1
 
         self.heartbeat_period = heartbeat_period
         self.comm = comm
         self.rank = rank
 
+    def create_reg_message(self):
```
[the remark this answers is not in the document]
Comment (yadudoc) [reply parsl/executors/extreme_scale/mpi_worker_pool.py]: Good catch! This is removed.

File: parsl/executors/high_throughput/process_worker_pool.py [hunk 8]
```diff
@@ -149,6 +174,9 @@ def pull_tasks(self, kill_event):
                         #    [i['task_id'] for i in self.pending_task_queue]))
             else:
                 logger.debug("[TASK_PULL_THREAD] No incoming tasks")
+                # Limit poll duration to heartbeat_period
```
[the remark this answers is not in the document]
Comment (yadudoc) [reply parsl/executors/high_throughput/process_worker_pool.py]: the current logic doubles the poll_timer *iff* the poll_timer after doubling is less than self.heartbeat_period. Perhaps the following is clearer : 

`poll_timer = min(self.heartbeat_period * 1000, poll_timer * 2)`

Also I just noticed that there's a units issue here, heartbeat_period is in seconds vs poll_timer in ms.

File: parsl/executors/extreme_scale/mpi_worker_pool.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (yadudoc) [reply parsl/executors/extreme_scale/mpi_worker_pool.py]: I tried to remove this and realized that I'd have to move this somewhere else for both executors. I've added a new issue to remind us of this: #644

File: parsl/executors/high_throughput/process_worker_pool.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (annawoodard) [reply parsl/executors/high_throughput/process_worker_pool.py]: Sorry, I misunderstood what you were trying to do. Thanks for the clarification.

**Pull request merged**